### PR TITLE
Add interface for cosmosdb MI

### PR DESCRIPTION
**Pull request by rbans96**
## Description
Describe the changes in this PR.

## Related issues
Addresses [workitem](https://microsofthealth.visualstudio.com/Health/_workitems/edit/131118)

## Testing
Describe how this change was tested.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Web/Startup.cs [hunk 1]
```diff
@@ -121,6 +121,7 @@ private void AddDataStore(IServiceCollection services, IFhirServerBuilder fhirSe
             if (runtimeConfiguration is AzureApiForFhirRuntimeConfiguration)
             {
                 fhirServerBuilder.AddCosmosDb();
+                fhirServerBuilder.AddCosmosdbMI();
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Web/Startup.cs:124]: Why not always register it, there is a separate config setting controlling this also

File: src/Microsoft.Health.Fhir.Core/Features/Operations/ICosmosDBAccessTokenProvider.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations
+{
+    public interface ICosmosDBAccessTokenProvider
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/ICosmosDBAccessTokenProvider.cs:12]: Should this also provide the TokenCredential to the FhirCosmosClientInitializer?

File: src/Microsoft.Health.Fhir.Shared.Web/Startup.cs (hunk 1 above)
Comment (rbans96) [reply src/Microsoft.Health.Fhir.Shared.Web/Startup.cs]: Updated!

File: src/Microsoft.Health.Fhir.Core/Features/Operations/ICosmosDBAccessTokenProvider.cs (hunk 2 above)
Comment (rbans96) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/ICosmosDBAccessTokenProvider.cs:14]: Very good point, updated the interface and FhirCosmosClientInitializer

File: src/Microsoft.Health.Fhir.CosmosDb/Registration/FhirServerBuilderCosmosDbRegistrationExtensions.cs [hunk 3]
```diff
@@ -271,6 +272,10 @@ private static IFhirServerBuilder AddCosmosDbPersistence(this IFhirServerBuilder
                 .AsImplementedInterfaces()
                 .AsFactory<IScoped<IQueueClient>>();
 
+            services.Add<CosmosAccessTokenProviderFactory>(c => c.GetRequiredService<IAccessTokenProvider>)
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.CosmosDb/Registration/FhirServerBuilderCosmosDbRegistrationExtensions.cs:275]: I think we need one more change for oss, we probably need to always register AzureAccessTokenProvider else this won't work if storage isn't setup

https://github.com/microsoft/fhir-server/blob/e891ceafb2f967110cb4df846945736fcbe3d1b5/src/Microsoft.Health.Fhir.Azure/FhirServerBuilderAzureRegistrationExtensions.cs#L49-L58

**Review by apurvabhaleMS — APPROVED** [review]

**Pull request merged**